### PR TITLE
Decouple `TokenTransfer*` and `SendPacket*`

### LINKED ISSUE #786: Decouple `TokenTransfer{Validation,Execution}` from `SendPacket{Validation,Execution}`
They are currently [coupled](https://github.com/cosmos/ibc-rs/blob/d6da3ed917a41cea1711bc2905dba8b4bcf6696c/crates/ibc/src/applications/transfer/context.rs#L29). However, basecoin shows us that this no longer works, since the `TransferModule` must know how to send packets, which now requires [access to the client validation context](https://github.com/cosmos/ibc-rs/blob/d6da3ed917a41cea1711bc2905dba8b4bcf6696c/crates/ibc/src/core/ics04_channel/context.rs#L30) (implemented by the core `Ibc` struct).

Basically, decoupling them just gives us more flexibility; they never *needed* to be coupled, it was just convenient.

### PR DESCRIPTION
**Pull request by plafer**
Closes: #786

______

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests.
- [ ] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: crates/ibc/src/mock/context/applications.rs [hunk 1]
```diff
@@ -11,7 +10,9 @@ use crate::applications::transfer::PrefixedCoin;
 use crate::core::ics24_host::identifier::{ChannelId, PortId};
 use crate::Signer;
 
-impl TokenTransferValidationContext for MockContext {
+pub struct MockTokenTransferModule;
+
+impl TokenTransferValidationContext for MockTokenTransferModule {
```
Comment (Farhad-Shabani) [inline crates/ibc/src/mock/context/applications.rs]: Following our discussion on Module vs Application, it makes more sense if `TokenTrafer*` contexts be implemented on the same struct that implements `Module`. The same as in [basecoin](https://github.com/informalsystems/basecoin-rs/blob/main/src/modules/ibc/transfer.rs#L122-L487).
We can impl them for `DummyTransferModule` under the `test_utils.rs`

File: crates/ibc/src/applications/transfer/context.rs [hunk 2]
```diff
@@ -444,6 +441,7 @@ pub(crate) mod test {
         let counterparty = Counterparty::new(port_id.clone(), Some(channel_id.clone()));
 
         (
+            MockTokenTransferModule,
             ctx,
```
Comment (Farhad-Shabani) [inline crates/ibc/src/applications/transfer/context.rs]: `ctx` not used anymore. Can entirely be removed.

File: crates/ibc/src/mock/context/applications.rs (hunk 1 above)
Comment (plafer) [reply crates/ibc/src/mock/context/applications.rs]: [93d6544](https://github.com/cosmos/ibc-rs/pull/787/commits/93d6544ce2b5ad28683b4fc7ff1dccacc4ec5eca)

File: crates/ibc/src/applications/transfer/context.rs (hunk 2 above)
Comment (plafer) [reply crates/ibc/src/applications/transfer/context.rs]: [34f658d](https://github.com/cosmos/ibc-rs/pull/787/commits/34f658db4df71861a4af853e611abf0bf78f8f0a)

**Review by Farhad-Shabani — APPROVED** [review]
🙏🏻

**Pull request merged**